Us 2.1: Show Events as List Discover pages

r-saba opened the pull request:
Closes #65 

To properly test this PR you will need to setup your database such that User with id 1

- Is not involved in at least 1 event
- Is a participant in at least 1 event, meaning they're in the `event_participants` table
- Is a host or coHost of at least 1 event, meaning in the `hostId` column there is a value of `1` and/or they're in the `event_cohosts` table

**Notes**
run `npm i`, I updated Apollo Client to the latest version to fix a bug with testing.

Images and Location on the event-card are hardcoded. Our API doesn't return this information yet.

**Additional Notes/Addressing PR changes**
I ensured the queries are always fetching the most up to date list of events. After creating an event there is a warning about updating state on an umounted component. [This is a known issue with Apollo client, no solution has been merged yet](https://github.com/apollographql/apollo-client/issues/6209). I don't see any performance or application issues as a result of this warning

r-saba committed:
feat(mull-ui): add string type to startDate and endDate in IEvent

the GraphQL query returns the datetime in string format, this was causing issues since our front-end
was expecting a Date Object.

Refs #65

r-saba committed:
feat(mull-ui): add list of events on dicover, upcoming and my-events tab of home page

Refs #65

r-saba committed:
refactor(mull-ui): move home page subpages into home directory

allowed me to put all repeated styles into home.discover.scss, reducing repeated code

Refs #65

r-saba committed:
fix(swipeable-routes): dynamically set height of swipeable route container based on content/screen

the height of the swipeable container would be the height of the route with the most content. This
allowed for scrolling on routes even when there was no content to show. In this commit the height of
hidden routes is set to 0 while the height of the visible route is set to 100%. This means the
height of the container will always equal the height of the content being shown or the height of the screen.

Refs #65

r-saba committed:
build(apollo-client): upgrade apollo-client version to latest

this fixes an issue when mocking graphql queries in front-end tests

Refs #65

r-saba committed:
test(mull-ui): update tests to use the new start and end date in the dummy event

Refs #65

r-saba committed:
test(mull-ui): add tests for home page subpages

Refs #65

r-saba committed:
feat(mull-ui): update discover page logic to show event cards when graphql data retrieved

Refs #65

r-saba committed:
feat(mull-ui): add typscript interfaces and return type for useQuery hooks in home components

Refs #65

r-saba committed:
style(event-card.tsx): pass the formatDate utility a date Object instead of datetime string

keeps the utility function doing one thing and one thing well

Refs #65

r-saba committed:
test(us-2.1.spec.ts): add e2e tests for US-2.1

Refs #65

r-saba committed:
test(mull-ui): fix tests that broke due to previous commits

Refs #65

r-saba committed:
refactor(mull-ui): remove unecessary casting on result of queries in discover tabs

after explicitly definig the return type of the useQuery hook I no longer needed to explictly cast
the data as Partial<IEvent>[]

Refs #65

r-saba committed:
feat(mull-ui): hide scrollbar from showing when swiping/changing between routes on home page

Refs #65

r-saba committed:
docs(utilities.ts): fix typo

Refs #65

cindyslittleplanet commented:
The tabs (the words) are too small compared to the buttons in the nav bar for the desktop view.
They are currently at size 16, which is the same size as the headers of each field to complete on the Create Event page. 

![image](https://user-images.githubusercontent.com/38992971/102925684-dd490a00-4461-11eb-8dce-76e1cc8f507a.png)

leojrsilao commented on apps/mull-ui/src/app/pages/home/home-discover.scss:
```diff
@@ -0,0 +1,14 @@
+.react-swipeable-view-container > div[aria-hidden='true'] {
+  max-height: calc(100vh - 11.5rem);
+  overflow-y: hidden !important; // hides scroll bar when changing routes. !important needed to overwrite the element style set by react swipeable views package.
```
I don't see the y-scrollbar without ```overflow-y: hidden !important``` when changing from the discover route to upcoming route or my events route unless it happens in a certain browser, I'm using google chrome. Also, there's probably a way to avoid using !important and override the react swipeable view package by using a class name or some property

r-saba replied:
Did you have enough events to scroll through? It did make a difference for me.
As for the !important, I needed it because there's inline style on the component and this can only be overridden by classes when you use `!important`

This is one of the acceptable cases to use `!important`, [here's some documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) on it, checkout out the "How !important can be used:" section

cindyslittleplanet reviewed (commented):
The tabs (the words) are too small compared to the buttons in the nav bar for the desktop view.
They are currently at size 16, which is the same size as the headers of each field to complete on the Create Event page.
And the space between the top and bottom of the tabs is not even. 
![image](https://user-images.githubusercontent.com/38992971/102935345-f8bd1080-4473-11eb-9395-954a711992f9.png)

leojrsilao reviewed (dismissed):
Good work! LGTM, the data is now queried for discovering events, upcoming events, and my events. 🚀 Just had a minor comment.

cindyslittleplanet reviewed (dismissed):
Since the issue I brought up can be fixed in PR #125 and was not part of the requirements of this US, this looks good to me now! 

- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
- [x] There must exist an automated Cypress acceptance test in GIF form
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/?node-id=533%3A6528)

RGPosadas reviewed (commented):
Everything currently works well, nice job! 👍 
The only issue I saw was that you would still need to do a manual browser refresh in order to get the updates from the db ([example vid](https://drive.google.com/file/d/1OK_M9o7vpIdd1_zl31gtZBTFUz1V3NRN/view?usp=sharing)). Does this happen to anyone else as well?

There's also a linting error on your E2E test.

r-saba committed:
chore(us-2.1.spec.ts): remove assignment of cypress returns to satisfy the linter

Refs #65

RGPosadas reviewed (changes requested):
Blocking due to what I wrote above.

cristian-aldea commented on apps/mull-ui/src/app/app.spec.tsx:
```diff
@@ -1,16 +1,18 @@
-      <BrowserRouter>
-        <App />
-      </BrowserRouter>
+      <MockedProvider>
```
Just a question. what does `MockedProvider` do?

r-saba replied:
It mocks the `ApolloProvider` used in `main.tsx`

TheGreatMarkus commented:
@RGPosadas yes this happens to me too, @r-saba what do you think about this issue?

cristian-aldea commented on apps/mull-ui/src/app/components/event-card/event-card.tsx:
```diff
@@ -22,7 +22,7 @@ export const EventCard = ({ event, style = {} }: EventCardProps) => {
   const distance = 15;
 
-  const { day, month, time } = formatDate(event.startDate);
+  const { day, month, time } = formatDate(new Date(event.startDate));
```
I ran into this issue too where `startDate` and `endDate` aren't really dates but are rather strings. Maybe the event service should be tweaked to fix this, so that querying for dates doesn't give back strings. What do you think @r-saba ?

r-saba replied:
The API gives us a serialized object and it's the responsibility of the front-end to deserialize it. This is why I added the `Date | string` type to the interface below https://github.com/RGPosadas/Mull/pull/129#discussion_r548065726. 

Let's continue this conversation below

cristian-aldea commented on apps/mull-ui/src/app/pages/home/discover/discover.tsx:
```diff
@@ -0,0 +1,36 @@
+  }
+`;
+
+export const DiscoverPage = () => {
```
`discover.(tsx|spec.tsx|scss)` files should be renamed to better fit the `DiscoverPage` name. Something like `discover-page`

cristian-aldea commented on apps/mull-ui/src/app/pages/home/my-events/my-events.tsx:
```diff
@@ -0,0 +1,44 @@
+
+export const MyEventsPage = () => {
+  const { data } = useQuery<MyEventData>(GET_PARTICIPATING_EVENTS, {
+    variables: { UserId: 1 },
```
Make sure to leave a `//TODO` for this, and all other places where the user id is hardcoded right now

cristian-aldea commented on libs/types/src/lib/types.ts:
```diff
@@ -31,8 +31,8 @@ export const EventRestrictionMap = ['Everyone', 'Friends', 'Invite Only'];
   title: string;
-  startDate: Date;
-  endDate: Date;
+  startDate: Date | string;
+  endDate: Date | string;
```
Do you really think this is a good idea? I'd rather not have to deal with composites types. It lead to some messy code in minicap.

r-saba replied:
Agreed, this was the one part I wasn't sure about. An option I was considering is another interface such as `ISerializedEvent` to store the returned data and then once we seralize the event we can use `IEvent`. What do you think, any suggestions?

r-saba replied:
I addressed this by adding a separate `ISeralizedEvent` interface to be used for the serialized event. I think this is the best solution since apollo client does not support automatic reverse seralization at the moment as per their [docs](https://www.apollographql.com/docs/apollo-server/schema/scalars-enums/#custom-scalars).

cristian-aldea replied:
I'm fine with this. I realized that my suggestion to update the IEvent interface to have `startDate` and `endDate` as strings might not be possible. Can you confirm that it's not possible, or did you prefer to have a separate interface for this?

r-saba replied:
Yes exactly, the `IEvent` is used in our API so I had to make a separate interface for our front-end

cristian-aldea replied:
Cool, sorry for being so insistent on that solution despite it not really being feasible haha.

cristian-aldea reviewed (changes requested):
Everything works great!

- [X] User Acceptance Flow is respected
- [X] Documentation present where necessary
- [X] Unit and UI tests are implemented for every component and function
- [X] There must exist an automated Cypress acceptance test in GIF form
- [X] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/?node-id=533%3A6528)

However, I had a few small blocking concerns:
* Please fix the linting issues you introduced, there's one in `event-card.tsx` and one in `US-2.1.spec.ts`
* you made it so that dates can either be `string` or `Date` types. And I just wanted to have a discussion about this before this is merged. I left a comment about that so you can reply to that comment.

r-saba committed:
feat(mull-ui): fetch events on home pages through network only

ensures the list of events is not cached and always up to date

Refs #65

r-saba committed:
docs(mull-ui): add TODO's for passing userId dynamically to queries in home pages

Refs #65

r-saba committed:
refactor(discover-page): rename discover.* to discover-page.*

Refs #65

r-saba committed:
style(mull-ui): remove unused variable from event-card.tsx

Refs #65

r-saba committed:
feat(mull-ui): add ISerializedEvent interface

apollo client does not support automatic reverse serialization. This solution allows us to perform
seralization when needed.

Refs #65

cristian-aldea reviewed (dismissed):
Thanks for addressing my comments. Everything looks good to me.

RGPosadas reviewed (dismissed):
Great job, LGTM! 🚀 

- [X] User Acceptance Flow is respected
- [X] Documentation present where necessary
- [X] Unit and UI tests are implemented for every component and function
- [X] There must exist an automated Cypress acceptance test in GIF form
- [X] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/?node-id=533%3A6528)

r-saba committed:
feat(mull-ui): pass onClick to event-card on home pages

Refs #65

r-saba committed:
style(mull-ui): rename css class match changes made to the name in develop

Refs #65

r-saba committed:
test: update tests to use seralized dates from mock events

Refs #65

r-saba committed:
refactor(mull-ui): remove unused scss files

Refs #65

RGPosadas reviewed (changes requested):
Blocking due to the failing unit tests on the discovery page sub-routes. Otherwise, linting, e2e, and pre-commit are all clean!

Screenshot of one of the tests:
![image](https://user-images.githubusercontent.com/33991121/103585796-beab3000-4eb1-11eb-980b-b9175a373db1.png)

RGPosadas reviewed (approved):
I forgot to run `npm install` so I didn't have the package that the unit test needed, so my previous blocking comment is irrelevant. 
LGTM!

cindyslittleplanet reviewed (approved):
Still looking good to me!
